File: 2026-02-February-GFG-POTD/February 10 - Koko Eating Bananas/Solution.cpp
class Solution {
public:
    int kokoEat(vector<int>& arr, int k) {
        int low = 1;
        int high = 0;

        // Find max pile
        for (int bananas : arr) {
            high = max(high, bananas);
        }

        int answer = high;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (canFinish(arr, k, mid)) {
                answer = mid;        // try smaller speed
                high = mid - 1;
            } else {
                low = mid + 1;       // need higher speed
            }
        }

        return answer;
    }

private:
    bool canFinish(vector<int>& arr, int k, int speed) {
        long long hours = 0;

        for (int bananas : arr) {
            hours += (bananas + speed - 1) / speed; // ceil division
            if (hours > k) return false;
        }

        return true;
    }
};

File: 2026-02-February-GFG-POTD/February 10 - Koko Eating Bananas/Solution.py
class Solution:
    def kokoEat(self, arr, k):
        low, high = 1, max(arr)
        answer = high

        while low <= high:
            mid = low + (high - low) // 2

            if self.canFinish(arr, k, mid):
                answer = mid      # try smaller speed
                high = mid - 1
            else:
                low = mid + 1     # need higher speed

        return answer

    def canFinish(self, arr, k, speed):
        hours = 0

        for bananas in arr:
            hours += (bananas + speed - 1) // speed  # ceil division
            if hours > k:
                return False

        return True
